Initial structure

Marenz opened the pull request:
- Prepare dependencies for dispatch
- Fix labeler path
- Add initial structure

Marenz committed:
Prepare dependencies for dispatch

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Fix labeler path

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Add highlevel dispatch interface

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Change structure to frequenz.dispatch

For this repository, we decided to apply the structure rewrite outlined in the following
document:

https://github.com/frequenz-floss/frequenz-sdk-python/wiki/APIs-stack-and-repositories-structure#completely-rewrite-of-the-structure

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

llucax commented on src/frequenz/dispatch/__init__.py:
````diff
@@ -0,0 +1,72 @@
+    This class provides a highlevel interface to the dispatch API. It
+    allows to receive new dispatches and ready dispatches.
+
+    Here is an example of how to use the `Dispatcher` class:
+
+    ```python
+    from frequenz.dispatch import Dispatcher
+
+    async def run():
+        dispatcher = Dispatcher(API_CONNECTION_INFO)
+        dispatcher.start()  # this will start the actor
+        dispatch_arrived = dispatcher.new_dispatches().new_receiver()
+        dispatch_ready = dispatcher.ready_dispatches().new_receiver()
+    ```
````
```suggestion
    Example:
        ```python
        from frequenz.dispatch import Dispatcher

        async def run():
            dispatcher = Dispatcher(API_CONNECTION_INFO)
            dispatcher.start()  # this will start the actor
            dispatch_arrived = dispatcher.new_dispatches().new_receiver()
            dispatch_ready = dispatcher.ready_dispatches().new_receiver()
        ```
```

Marenz committed:
Ci: Reuse nox environment for crossbuilds

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

shsms commented:
Could you describe the motivation for this change?  There's `from frequenz import dispatch` now?  Aren't we only having namespace packages under frequenz like `frequenz.{actor, client, api, etc}` and putting code under that?

Or is this going to be a completely independent thing like `frequenz.channels`?

Marenz commented:
The structure is a kind of experiment that luca suggested to me to try. 
Basically, as the user shouldn't interact directly with the actor, the high-level interface should just directly offer the dispatch arrived/is ready channels like shown in https://github.com/frequenz-io/frequenz-apps-draft/blob/main/dispatch_exec/dispatch.py

shsms commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,199 @@
+        """Run the actor."""
+        try:
+            while True:
+                # for _ in range(30):
```
I guess this can go away.

shsms commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,199 @@
+            while True:
+                # for _ in range(30):
+                await self._fetch()
+                await asyncio.sleep(timedelta(hours=5).total_seconds())
```
maybe you want to take the 5 hours from a constructor argument with a default value.  That way you can give it a name and a description, so people would know why 5 hours.

Marenz replied:
I meant to use the constant (which is also missing docs) `_MAX_AHEAD_SCHEDULE`

shsms commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,199 @@
+        self._dispatches = {}
+
+        try:
+            logging.info("Fetching dispatches for microgrid %s", self._microgrid_id)
```
I think we should always have `_logger = logging.getLogger(__name__)` at the top of the file and use `_logger.info`, etc. everywhere.

With that, we'll be able to do filter levels like `logging.getLogger("frequenz.actor").setLevel(logging.DEBUG)`, and I think we should support that in all our packages.

shsms replied:
Maybe we should look into making a pylint plugin for that.

shsms commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,199 @@
+            await self._ready_dispatch_sender.send(dispatch)
+
+    @classmethod
+    def calculate_next_run(cls, dispatch: Dispatch, _from: datetime) -> datetime | None:
```
why does this have to be a public class method?  from what I can see, it will be used only from within the `Dispatcher`, and only from contexts where a `self` is available.

Marenz replied:
It doesn't have to be, but as a user who wants to create new dispatches and investigate existing ones, I figured it's very useful to know when the next time is a dispatch would run.

shsms commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,199 @@
+                continue
+
+            logging.info("Dispatch %s scheduled for %s", dispatch.id, next_time)
+            logging.info(
+                "Sleeping for %s",
+                next_time - datetime.now().replace(tzinfo=timezone.utc),
+            )
+            await asyncio.sleep(
+                (
+                    next_time - datetime.now().replace(tzinfo=timezone.utc)
+                ).total_seconds()
+            )
```
maybe we can reuse the `now()` values, because it is a separate syscall each time.

shsms replied:
also,  isn't this the same as `datetime.now(tz=timezone.utc)`?  Or does `replace` have a different behaviour?

Marenz replied:
it's pretty much the same

shsms commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,199 @@
+        Args:
+            dispatch: The dispatch to schedule.
+        """
+        while True:
+            next_time = self.calculate_next_run(
+                dispatch, datetime.now().replace(tzinfo=timezone.utc)
+            )
+            if next_time is None:
+                logging.info("Dispatch finished: %s", dispatch)
+                self._scheduled.pop(dispatch.id)
+                return
```
maybe more readable with:

```python
while next_time := self.calculate_next_run(dispatch, datetime.now(tz=timezone.utc)):
    ...

# and after the loop
logging.info("Dispatch finished: %s", dispatch)
self._scheduled.pop(dispatch.id)
```

Marenz replied:
then I can't reuse `now` easily with your suggestion, but I modified it a bit..

```python
        def next_run_info() -> tuple[datetime, datetime] | None:
            now = datetime.now(tz=timezone.utc)
            next_run = self.calculate_next_run(dispatch, now)

            if next_run is None:
                return None

            return now, next_run

        while pair := next_run_info():
            now, next_time = pair

            if next_time - now > _MAX_AHEAD_SCHEDULE:
                await asyncio.sleep(_MAX_AHEAD_SCHEDULE.total_seconds())
                continue

            _logger.info("Dispatch %s scheduled for %s", dispatch.id, next_time)
            _logger.info("Sleeping for %s", next_time - now)
            await asyncio.sleep((next_time - now).total_seconds())

            _logger.info("Dispatch ready: %s", dispatch)
            await self._ready_dispatch_sender.send(dispatch)

        _logger.info("Dispatch finished: %s", dispatch)
        self._scheduled.pop(dispatch.id)

```

shsms commented:
> The structure is a kind of experiment that luca suggested to me to try.
> Basically, as the user shouldn't interact directly with the actor, the high-level interface should just directly offer the dispatch arrived/is ready channels like shown in https://github.com/frequenz-io/frequenz-apps-draft/blob/main/dispatch_exec/dispatch.py

Then maybe we keep `frequenz.actor.dispatch` as it is, and add `frequenz.dispatch` separately, such that `frequenz.dispatch` can eventually go into the SDK?

Either way, the reason for the change is not obvious at all, so it would be worth the effort to explain it in the commit message and in the PR description.

Marenz commented:
I don't see it much as "change of structure", given that previously there was basically nothing except empty or dummy files  ;)

Anyway, I found what I was missing. The structure change is an attempt to test the rewrite of the structure according to https://github.com/frequenz-floss/frequenz-sdk-python/wiki/APIs-stack-and-repositories-structure#completely-rewrite-of-the-structure

Marenz commented:
* Added link to structure rewrite description in commit
* implemented proper scheduling with recurrence based on `python-dateutils`
* can use more tests, but that can be added later (test suggestions welcome, part from the create/delete/update tests that are there/planned)

Marenz committed:
Use new type DispatchEvent for new/updated/deleted dispatches

Also add some tests for it.

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

Marenz committed:
Make poll interval configurable

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

shsms commented on README.md:
```diff
@@ -15,22 +15,30 @@ It provides two channels for clients:
+                        ...
+                    case Deleted(dispatch):
+                        log.info("Dispatch deleted %s", dispatch.deleted)
+                        ...
```
I think you don't need the `.created`, etc. anymore.

shsms commented:
Looks good to me.

stefan-brus-frequenz commented on src/frequenz/dispatch/__init__.py:
```diff
@@ -0,0 +1,71 @@
+    """A highlevel interface for the dispatch API.
+
+    This class provides a highlevel interface to the dispatch API. It
+    allows to receive new dispatches and ready dispatches.
```
Suggestion:

```
This class provides a highlevel interface to the dispatch API. It 
allows for receiving of new dispatches, and dispatches labelled as 'ready'.
```

I would also suggest adding a small blurb about the difference between the two.

stefan-brus-frequenz commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,275 @@
+    ) -> None:
+        """Update the schedule for a dispatch.
+
+        Schedules, reschedules or cancels the dispatch based on the start_time
```
Could you clarify how this happens in this doc?

For example, "- If the start time of the new dispatch is different from the old one, the old one is cancelled."

stefan-brus-frequenz reviewed (dismissed):
Added a couple of optional comments about the docs, code overall LGTM, but my python knowledge is limited.

Marenz committed:
Explain functionality more in the documentation

Signed-off-by: Mathias L. Baumann <mathias.baumann@frequenz.com>

llucax commented on pyproject.toml:
```diff
@@ -64,11 +69,13 @@ dev-mkdocs = [
 ]
 dev-noxfile = [
-  "nox == 2023.4.22",
+  "uv == 0.1.14",
```
:scream:

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/issues/13

llucax commented on pyproject.toml:
```diff
@@ -26,21 +25,27 @@ classifiers = [
-  "frequenz-sdk == 0.25.2",
+  "frequenz-sdk == v1.0.0-rc5",
+  "frequenz-channels == 1.0.0b2",
+  "frequenz-dispatch-client @ git+https://github.com/frequenz-floss/frequenz-client-dispatch-python.git@00dcb3c",
```
Oh, damn, this python package name is wrong, it should be `frequenz-client-dispatch`).

* https://github.com/frequenz-floss/frequenz-client-dispatch-python/pull/27

llucax commented on pyproject.toml:
```diff
@@ -26,21 +25,27 @@ classifiers = [
+  "frequenz-sdk == v1.0.0-rc5",
+  "frequenz-channels == 1.0.0b2",
+  "frequenz-dispatch-client @ git+https://github.com/frequenz-floss/frequenz-client-dispatch-python.git@00dcb3c",
+  "frequenz-api-dispatch @ git+https://github.com/frequenz-floss/frequenz-api-dispatch.git@223315c",
+#  "frequenz-api-dispatch >= 0.13.0, < 0.14",
+  "frequenz-client-base >= 0.2.1, < 0.3.0",
+  # Directly use unreleased commit until the first release
+  "frequenz-client-common @ git+https://github.com/frequenz-floss/frequenz-client-common-python.git@5e6e7b7",
```
These 3 are here because of `uv`? Or are they really used directly by this project? And if so, why, intuitively this project should only depend on `client-dispatch`.

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/issues/13

llucax commented on src/frequenz/dispatch/__init__.py:
```diff
@@ -0,0 +1,78 @@
+
+from frequenz.dispatch.actor import DispatchActor, DispatchEvent
+
+__all__ = ["Dispatcher"]
```
We need to export more stuff here, all that's returned by the public interface too, like `DispatchEvent` and all the stuff that's exposed by it.

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax commented on src/frequenz/dispatch/__init__.py:
```diff
@@ -0,0 +1,78 @@
+__all__ = ["Dispatcher"]
+
+
+class Dispatcher:
```
Since we normally don't put actual code in `__init__.py`, I suggest moving this to `_dispatcher.py`.

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,279 @@
+from frequenz.sdk.actor import Actor
+
+_MAX_AHEAD_SCHEDULE = timedelta(hours=5)
+"""The maximum time ahead to schedule a dispatch.
+
+We don't want to schedule dispatches too far ahead,
+as they could start drifting if the delay is too long.
+
+This also prevents us from scheduling too many dispatches at once.
+
+The exact value is not important, but should be a few hours and not more than a day.
+"""
```
I didn't understand which issue this is trying to solve, maybe we need a more detailed explanation of the issue and how this solves it.

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/issues/16

llucax commented on src/frequenz/dispatch/__init__.py:
```diff
@@ -0,0 +1,78 @@
+        """Start the actor."""
+        self._actor.start()
+
+    def updated_dispatches(self) -> Receiver[DispatchEvent]:
```
Now that this handles create and delete too, maybe `updated` is not the most clear choice. When I described the issue, ChatGPT suggested `dispatch_lifecycle_events` and `ready_for_execution_dispatches` but it seems a bit too long. I wonder if we really need to add `dispatch` to the name, as it should be pretty obvious in this object's context.

Maybe just `dispatcher.lifecycle_events` and `dispatcher.ready_to_execute`?

Another approach would be to unify both channels and just have `dispatcher.dispatches`, but I guess many users will just want to just subscribe to the ready dispatches and all the rest will be (a considerable amount of) noise.

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax commented on src/frequenz/dispatch/__init__.py:
```diff
@@ -0,0 +1,78 @@
+        Returns:
+            A new receiver for new dispatches.
+        """
+        return self._updated_channel.new_receiver()
```
For the SDK we decided to call stuff creating new objects more clearly, as people might think that a receiver is stored internally. I think we should do the same here, so it should be something like: `dispatcher.lifecycle_events.new_receiver()` and `dispatcher.ready_to_execute.new_receiver()`? (following the suggestion above)

We can just return the internal channel if we change the return type to a [`ReceiverFetcher`](https://github.com/frequenz-floss/frequenz-sdk-python/blob/1a3dce7623684120c999ff23b86fa984a4890471/src/frequenz/sdk/_internal/_channels.py#L15-L27) protocol (we can just copy&paste the protocol, it is really super lightweight.

So the implementation can simply be:

```py
    @property
    def lifecycle_events(self) -> ReceiverFetcher[DispatchEvent]:
        return self._updated_channel
```

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax commented on src/frequenz/dispatch/__init__.py:
```diff
@@ -0,0 +1,78 @@
+        """
+        return self._updated_channel.new_receiver()
+
+    def ready_dispatches(self) -> Receiver[Dispatch]:
```
Same here.

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,279 @@
+
+
+@dataclass(frozen=True)
+class _DispatchEventBase:
```
This should be public (`DispatchEventBase`) if public classes inherit from it. But I think, being just a super thin wrapper, we should just not inherit from a common base, which also could lead to user mistakes if they use the base class in a `match` statement.

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,279 @@
+"""
+
+
+class DispatchActor(Actor):
```
Technically this should be named `DispatchingActor` according to the naming convention :grimacing:

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,279 @@
+        try:
+            while True:
+                await self._fetch()
+                await asyncio.sleep(self._poll_interval.total_seconds())
```
We should use a `Timer` here to properly deal with drift, even when unlikely (it will make the code simpler too):
`async for drift in self._poll_timer: await self._fetch()` (we could also print the drift or something).

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/pull/15

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,279 @@
+            while True:
+                await self._fetch()
+                await asyncio.sleep(self._poll_interval.total_seconds())
+        except asyncio.CancelledError:
+            for task in self._scheduled.values():
+                task.cancel()
+            raise
```
We need https://github.com/frequenz-floss/frequenz-sdk-python/issues/819!

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,279 @@
+            while True:
+                await self._fetch()
+                await asyncio.sleep(self._poll_interval.total_seconds())
+        except asyncio.CancelledError:
+            for task in self._scheduled.values():
+                task.cancel()
+            raise
```
It took me a while to think about the different scenarios to conclude calling cancel only when `CancelledError` is raised. Just to confirm my thinking is correct, you do this because if any other unexpected exception is raised and `_run()` is aborted, the auto-restart for actors will kick-in and meanwhile, any task that it is still healthy, will continue working, so hopefully ready dispatches will still be delivered on time, even if the actor wasn't restarted yet?

If this is the case, maybe you should check tasks health when `_fetch()`ing too, in case some task was aborted, to try to restart it. Right now you are only checking if a task is in the `_scheduled` dict, but not if the tasks are still running or not.

In any case, it is probably worth adding a comment here clarifying how clean up works exactly on different scenarios.

llucax replied:
* https://github.com/frequenz-floss/frequenz-dispatch-python/issues/17

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,279 @@
+                self._schedule_task(dispatch)
+            )
+
+    async def _schedule_task(self, dispatch: Dispatch) -> None:
```
My mental picture of how this should have worked was, instead of having one task per dispatch, to just have one task that calculates the next run for all dispatches and task the one that needs to run next, then sleeps only for that time, sends the ready dispatch through the channel, and calculates the next dispatch to be run again.

I think this scheme should be simpler and more memory efficient, as you don't need one task per dispatch. Do you see any advantages with the current approach?

I guess it might be a bit more CPU efficient, as the next dispatch for all dispatches doesn't need to be calculated every time a dispatch is ready, but this could be optimized fairly easily by saving all the next run calculation for all dispatches in a sorted list and only calculate the next run for the dispatch that was just ready. Then we just need to invalidate this list each time dispatches are changed.

llucax replied:
This one might be too much work to implement, so I think it is better to discuss it first, and probably it is better to sense the general feeling here first, if it's not a trivial "let's go with this approach" (or let's NOT) before creating an issue, so please go ahead and leave comments here.

Marenz replied:
I think my only concern with your suggestion is that it might accidentally skip dispatches that are very close together, though I can't come up for a good explanation why it would, just an .. gut feeling to pay extra attention on that area when implementing it..

I was also wondering if the awaiting for sending might take a bit of time and thus delays the notification for the next very close dispatch to be executed. What if we waited longer than the next dispatches scheduled time, how do we make sure we don't skip it?

llucax replied:
> I think my only concern with your suggestion is that it might accidentally skip dispatches that are very close together, though I can't come up for a good explanation why it would, just an .. gut feeling to pay extra attention on that area when implementing it..

If done properly, I don't think it should. I didn't invented this, I've seen this (or a similar approach) suggested in [libev](http://software.schmorp.de/pkg/libev.html) as a [very efficient way to implement timers](http://pod.tst.eu/http://cvs.schmorp.de/libev/ev.pod#code_ev_timer_code_relative_and_opti) (see point 4).

> I was also wondering if the awaiting for sending might take a bit of time and thus delays the notification for the next very close dispatch to be executed. What if we waited longer than the next dispatches scheduled time, how do we make sure we don't skip it?

If we await on the send takes a long time, I guess the next notification will be delayed too as it will use the same channel to notify, right? We could do the send in a separate task but then we end up with many tasks again, but I think sends will be serialized in any case because they operate in the same channel, so using a different task is probably not having any real effect.

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,279 @@
+        self._scheduled.pop(dispatch.id)
+
+    @classmethod
+    def calculate_next_run(cls, dispatch: Dispatch, _from: datetime) -> datetime | None:
```
Agreed that is useful to have, I even wonder if we should move this method to the `Dispatch` class instead (so to the client), seems like something anyone using the API client directly probably want to have too, an easy way to know when a dispatch should be run next.

llucax replied:
* https://github.com/frequenz-floss/frequenz-client-dispatch-python/issues/28

llucax commented on src/frequenz/dispatch/actor.py:
```diff
@@ -0,0 +1,279 @@
+            interval=dispatch.recurrence.interval,
+        )
+
+        return cast(datetime | None, next_run.after(_from, inc=True))
```
Is this `cast` needed because `dateutil` lacks type info or what?

llucax replied:
This one is too small to create an issue. So also please comment here.

Marenz replied:
Yah, I was getting `src/frequenz/dispatch/actor.py:279: error: Returning Any from function declared to return "datetime | None"  [no-any-return]`

llucax replied:
OK, it is probably best to add a comment when using `cast`, so readers don't have to guess in the future.

llucax reviewed (commented):
OK, quite a few comments, in particular about the design.

> [!IMPORTANT]
> Please let's not discuss the topics here, as discussion could get lengthy. I will create separate issues (or PRs where possible) and link them when I'm done.
